lib/PuppeteerSharp.Tests/Page/EvaluateTests.cs

```diff
@@ -0,0 +1,127 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace PuppeteerSharp.Tests.Page
+{
+    [Collection("PuppeteerLoaderFixture collection")]
+    public class EvaluateTests : PuppeteerBaseTest
+    {
+        [Theory]
+        [InlineData("1 + 5;", 6)] //ShouldAcceptSemiColons
+        [InlineData("2 + 5\n// do some math!'", 7)] //ShouldAceptStringComments
+        public async Task BasicIntExressionEvaluationTest(string script, object expected)
+        {
+            using (var page = await Browser.NewPageAsync())
+            {
+                var result = await page.EvaluateExpressionAsync<int>(script);
+                Assert.Equal(expected, result);
+            }
+        }
+
+        [Theory]
+        [InlineData("() => 7 * 3", 21)] //ShouldWork
+        [InlineData("() => Promise.resolve(8 * 7)", 56)] //ShouldAwaitPromise
+        public async Task BasicIntFunctionEvaluationTest(string script, object expected)
+        {
+            using (var page = await Browser.NewPageAsync())
+            {
+                var result = await page.EvaluateFunctionAsync<int>(script);
+                Assert.Equal(expected, result);
+            }
+        }
+
+        [Fact]
+        public async Task ShouldWorkRightAfterFrameNavigated()
+        {
+            Task<int> frameEvaluation = null;
+
+            using (var page = await Browser.NewPageAsync())
+            {
+                page.FrameNavigated += (sender, e) =>
+                {
+                    frameEvaluation = e.Frame.EvaluateFunctionAsync<int>("() => 6 * 7");
+                };
+
+                await page.GoToAsync(TestConstants.EmptyPage);
+                Assert.Equal(42, await frameEvaluation);
+            }
+        }
+
+        [Fact]
+        public async Task ShouldRejectPromiseWithExeption()
+        {
+            using (var page = await Browser.NewPageAsync())
+            {
+                var exception = await Assert.ThrowsAsync<EvaluationFailedException>(() =>
+                {
+                    return page.EvaluateFunctionAsync<object>("() => not.existing.object.property");
+                });
+
+                Assert.NotNull(exception);
```

Comment:
pretty sure `exception` cannot be null

Reply:
Why?

Reply:
the previous line will throw if the returned type is not an `EvaluationFailedException`

Reply:
Gotcha!